### DISCUSSION TITLE
Python crashes after generating embedding

### DISCUSSION THREAD

**Discussion by a-meneghini** [MaartenGr/BERTopic#1562] [Q&A]
Hi there, amazing work here with BERTopic. I'm writing asking for some help. I'm encountering python kernel crash during the analysis of a quite big corpus. We are speaking of a dataframe with 1.929.520 rows, approx 3 gb of text data. Over a colab pro session with 50 gb of RAM and 15 gb of VRAM (T4 GPU), i manage to calculate the embeddings (so Batches: 100%) with little problems, but then python session crashes and colab says "Your session crashed after using all available RAM". Any workaround for this?
I am using bertopic-0.15.0
Thanks!

**Comment by MaartenGr** [comment]
In the [ReadMe](https://github.com/MaartenGr/BERTopic#getting-started) of BERTopic, there is a guide for training it on large data. Have you followed that guide?

**Reply by a-meneghini** [reply]
Hi, thanks a lot for the support and for linking to that guide that I didn't consult previously. However, now there's a new error after reducing dimensionality, looks like this is a problem with cuml. Here's the traceback:

> `[D] [07:55:15.953908] /__w/cuml/cuml/cpp/src/umap/runner.cuh:108 n_neighbors=50
> [D] [07:55:15.955450] /__w/cuml/cuml/cpp/src/umap/runner.cuh:130 Calling knn graph run
> [D] [08:12:07.405804] /__w/cuml/cuml/cpp/src/umap/runner.cuh:136 Done. Calling fuzzy simplicial set
> [D] [08:12:07.827283] /__w/cuml/cuml/cpp/src/umap/fuzzy_simpl_set/naive.cuh:317 Smooth kNN Distances
> [D] [08:12:07.832354] /__w/cuml/cuml/cpp/src/umap/fuzzy_simpl_set/naive.cuh:319 sigmas = [ 0.000276196, 0.174828, 0.0764434, 0.168949, 3.8147e-10, 7.34851e-05, 0.000243747, 0.000297125, 0.0626316, 0.0712512, 0.0508443, 7.62155e-08, 0.0412415, 0.0259095, 0.0173285, 1.03085e-05, 0.0292451, 0.217813, 0.000274574, 7.34852e-05, 0.0464972, 0.1188, 0.183553, 0.0233847, 0.0815217 ]
> 
> [D] [08:12:07.832468] /__w/cuml/cuml/cpp/src/umap/fuzzy_simpl_set/naive.cuh:321 rhos = [ 0.283717, 1.19209e-07, 5.96046e-08, 5.96046e-08, 2.98023e-07, 0.0149949, 0.180006, 0.31879, 0.179992, 0.302594, 0.150941, 1.78814e-07, 0.250383, 0.363263, 0.305736, 1.19209e-07, 0.381917, 5.96046e-08, 1.78814e-07, 0.0149949, 0.327972, 5.96046e-08, 1.19209e-07, 0.201553, 0.173059 ]
> 
> [D] [08:12:07.832536] /__w/cuml/cuml/cpp/src/umap/fuzzy_simpl_set/naive.cuh:345 Compute Membership Strength
> [D] [08:13:07.817096] /__w/cuml/cuml/cpp/src/umap/fuzzy_simpl_set/naive.cuh:348 /__w/cuml/cuml/cpp/src/umap/fuzzy_simpl_set/naive.cuh:348 
> [D] [08:13:09.090553] /__w/cuml/cuml/cpp/src/umap/runner.cuh:143 Done. Calling remove zeros
> [D] [08:14:35.271790] /__w/cuml/cuml/cpp/src/umap/simpl_set_embed/algo.cuh:347 /__w/cuml/cuml/cpp/src/umap/simpl_set_embed/algo.cuh:347 
> [D] [08:16:41.807979] /__w/cuml/cuml/cpp/src/umap/runner.cuh:347 Running transform
> [D] [08:16:41.808093] /__w/cuml/cuml/cpp/src/umap/runner.cuh:349 Building KNN Graph
> [D] [08:33:38.797666] /__w/cuml/cuml/cpp/src/umap/runner.cuh:382 Smoothing KNN distances
> [D] [08:33:39.152094] /__w/cuml/cuml/cpp/src/umap/runner.cuh:414 Executing fuzzy simplicial set
> [D] [08:33:39.210368] /__w/cuml/cuml/cpp/src/umap/runner.cuh:443 Performing L1 normalization
> [D] [08:33:39.861353] /__w/cuml/cuml/cpp/src/umap/runner.cuh:479 n_epochs=30
> 
> 2023-10-06 08:33:46,096 - BERTopic - Reduced dimensionality
> 
> [D] [08:33:40.270644] /__w/cuml/cuml/cpp/src/umap/runner.cuh:502 Computing # of epochs for training each sample
> [D] [08:33:40.274577] /__w/cuml/cuml/cpp/src/umap/runner.cuh:509 Performing optimization
> 
> ---------------------------------------------------------------------------
> 
> RuntimeError                              Traceback (most recent call last)
> 
> [<ipython-input-10-df993e2dc72f>](https://localhost:8080/#) in <cell line: 12>()
>      10 
>      11 # Fit BERTopic without actually performing any clustering
> ---> 12 topic_model= BERTopic(
>      13         embedding_model=embedding_model,
>      14         umap_model=umap_model,
> 
> 5 frames
> 
> [/usr/local/lib/python3.10/dist-packages/cuml/internals/api_decorators.py](https://localhost:8080/#) in wrapper(*args, **kwargs)
>     188                         ret = func(*args, **kwargs)
>     189                     else:
> --> 190                         return func(*args, **kwargs)
>     191 
>     192                 return cm.process_return(ret)
> 
> base.pyx in cuml.internals.base.UniversalBase.dispatch_func()
> 
> hdbscan.pyx in cuml.cluster.hdbscan.hdbscan.HDBSCAN.fit()
> 
> RuntimeError: CUDA error encountered at: file=/__w/cuml/cuml/python/_skbuild/linux-x86_64-3.10/cmake-build/_deps/raft-src/cpp/include/raft/util/cudart_utils.hpp line=174: call='cudaMemcpyAsync(d_ptr1, d_ptr2, len * sizeof(Type), cudaMemcpyDeviceToDevice, stream)', Reason=cudaErrorInvalidValue:invalid argument
> Obtained 64 stack frames
> #0 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(_ZN4raft9exception18collect_call_stackEv+0x81) [0x7807698d8401]
> #1 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(_ZN4raft10cuda_errorC1ERKSs+0x14b) [0x7807698d8e8b]
> #2 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(_ZN4raft10copy_asyncIiEEvPT_PKS1_mN3rmm16cuda_stream_viewE+0x185) [0x780769e1e975]
> #3 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(+0x2523579) [0x780769ec0579]
> #4 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(_ZN2ML7HDBSCAN13build_linkageIifEEvRKN4raft8handle_tEPKT0_mmNS2_8distance12DistanceTypeERNS0_6Common13HDBSCANParamsEPS6_RNSB_28robust_single_linkage_outputIT_S6_EE+0x7b8) [0x780769ec2798]
> #5 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(_ZN2ML7HDBSCAN12_fit_hdbscanIifEEvRKN4raft8handle_tEPKT0_mmNS2_8distance12DistanceTypeERNS0_6Common13HDBSCANParamsEPT_PS6_RNSB_14hdbscan_outputISE_S6_EE+0x110) [0x780769ec3070]
> #6 in /usr/local/lib/python3.10/dist-packages/cuml/internals/../libcuml++.so(_ZN2ML7hdbscanERKN4raft8handle_tEPKfmmNS0_8distance12DistanceTypeERNS_7HDBSCAN6Common13HDBSCANParamsERNS9_14hdbscan_outputIifEEPf+0x228) [0x780769e00ce8]
> #7 in /usr/local/lib/python3.10/dist-packages/cuml/cluster/hdbscan/hdbscan.cpython-310-x86_64-linux-gnu.so(+0x79a7a) [0x78089588ca7a]
> #8 in /usr/local/lib/python3.10/dist-packages/cuml/internals/base.cpython-310-x86_64-linux-gnu.so(+0x24ff8) [0x7809174afff8]
> #9 in /usr/bin/python3(PyObject_Call+0xbb) [0x5cae7900e12b]
> #10 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x2b71) [0x5cae78fea2c1]
> #11 in /usr/bin/python3(+0x16e4e1) [0x5cae7900d4e1]
> #12 in /usr/bin/python3(PyObject_Call+0x122) [0x5cae7900e192]
> #13 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x2b71) [0x5cae78fea2c1]
> #14 in /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x5cae78fff70c]
> #15 in /usr/bin/python3(PyObject_Call+0x122) [0x5cae7900e192]
> #16 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x2b71) [0x5cae78fea2c1]
> #17 in /usr/bin/python3(+0x16e4e1) [0x5cae7900d4e1]
> #18 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x1981) [0x5cae78fe90d1]
> #19 in /usr/bin/python3(+0x16e4e1) [0x5cae7900d4e1]
> #20 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x1981) [0x5cae78fe90d1]
> #21 in /usr/bin/python3(+0x16e4e1) [0x5cae7900d4e1]
> #22 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x1981) [0x5cae78fe90d1]
> #23 in /usr/bin/python3(+0x16e4e1) [0x5cae7900d4e1]
> #24 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x1981) [0x5cae78fe90d1]
> #25 in /usr/bin/python3(+0x239e56) [0x5cae790d8e56]
> #26 in /usr/bin/python3(PyEval_EvalCode+0x86) [0x5cae790d8cf6]
> #27 in /usr/bin/python3(+0x23fb0d) [0x5cae790deb0d]
> #28 in /usr/bin/python3(+0x160969) [0x5cae78fff969]
> #29 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x6bd) [0x5cae78fe7e0d]
> #30 in /usr/bin/python3(+0x17d880) [0x5cae7901c880]
> #31 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x2786) [0x5cae78fe9ed6]
> #32 in /usr/bin/python3(+0x17d880) [0x5cae7901c880]
> #33 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x2786) [0x5cae78fe9ed6]
> #34 in /usr/bin/python3(+0x17d880) [0x5cae7901c880]
> #35 in /usr/bin/python3(+0x25a13f) [0x5cae790f913f]
> #36 in /usr/bin/python3(+0x16bf8a) [0x5cae7900af8a]
> #37 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x802) [0x5cae78fe7f52]
> #38 in /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x5cae78fff70c]
> #39 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x6bd) [0x5cae78fe7e0d]
> #40 in /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x5cae78fff70c]
> #41 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x802) [0x5cae78fe7f52]
> #42 in /usr/bin/python3(+0x16e4e1) [0x5cae7900d4e1]
> #43 in /usr/bin/python3(PyObject_Call+0x122) [0x5cae7900e192]
> #44 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x2b71) [0x5cae78fea2c1]
> #45 in /usr/bin/python3(+0x16e4e1) [0x5cae7900d4e1]
> #46 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x1981) [0x5cae78fe90d1]
> #47 in /usr/bin/python3(+0x205f05) [0x5cae790a4f05]
> #48 in /usr/bin/python3(+0x160969) [0x5cae78fff969]
> #49 in /usr/bin/python3(+0x23b8a5) [0x5cae790da8a5]
> #50 in /usr/bin/python3(+0x2b52a2) [0x5cae791542a2]
> #51 in /usr/bin/python3(+0x15350b) [0x5cae78ff250b]
> #52 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x6bd) [0x5cae78fe7e0d]
> #53 in /usr/bin/python3(_PyFunction_Vectorcall+0x7c) [0x5cae78fff70c]
> #54 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x802) [0x5cae78fe7f52]
> #55 in /usr/bin/python3(+0x205f05) [0x5cae790a4f05]
> #56 in /usr/bin/python3(+0x160969) [0x5cae78fff969]
> #57 in /usr/bin/python3(+0x23b8a5) [0x5cae790da8a5]
> #58 in /usr/bin/python3(+0x2b52a2) [0x5cae791542a2]
> #59 in /usr/bin/python3(+0x15350b) [0x5cae78ff250b]
> #60 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x6bd) [0x5cae78fe7e0d]
> #61 in /usr/bin/python3(+0x16e4e1) [0x5cae7900d4e1]
> #62 in /usr/bin/python3(_PyEval_EvalFrameDefault+0x6bd) [0x5cae78fe7e0d]
> #63 in /usr/bin/python3(+0x205f05) [0x5cae790a4f05]`

Now i'm trying to launch the analysis without cuml and see what happens
EDIT: it crashes for exceeding the available RAM 😢

**Reply by a-meneghini** [reply]
Hi, somehow I managed to run properly my operations by forcing llvmlite to version 0.39.1
Thanks for the support and great documentation!
